[Peterborough][WW] Add text messaging option for bulky waste

MorayMySoc opened the pull request:
Adds text box to opt into text updates for bulky waste bookings.

Allows multiple govuk_notify templates/keys to be added in array


https://github.com/mysociety/societyworks/issues/3782

[skip changelog]

MorayMySoc committed:
[Peterborough][WW] Adds text message request checkbox to Bulky

Adds text message request checkout to bulky waste details page
that is active only for Peterborough

https://github.com/mysociety/societyworks/issues/3782

MorayMySoc committed:
[FMS] Allow govuk_notify config to be array

Checks if govuk_notify is an array which will contain
hashes that now include type.

If the config is not an array continues as previously,
otherwise will choose which config to use based on
data assigned to notify_choice

Related to https://github.com/mysociety/societyworks/issues/3782

MorayMySoc committed:
[Peterborough][WW] Optionally send bulky waste text updates

If the user chose to receive updates by phone when making their
bulky waste booking, send text updates for 3 and 1 day reminders.

Continue to send emails if there is an email address.

https://github.com/mysociety/societyworks/issues/3782

MorayMySoc committed:
[Peterborough][WW] Send optional text on bulky waste confirmation

If the user has elected to receive optional texts for updates
to their booking, they will be sent a text when the bulky
booking is confirmed.

Normal email confirmation also takes place.

https://github.com/mysociety/societyworks/issues/3782

MorayMySoc committed:
[Peterborough][WW] Add tests for text sending

Adds tests for optional text messaging for Bulky waste bookings

https://github.com/mysociety/societyworks/issues/3782

davea commented on perllib/FixMyStreet/App/Form/Waste/Bulky.pm:
```diff
@@ -34,9 +34,15 @@ has_page cannot_book => (
     next => 'choose_date_earlier',
+    update_field_list => sub {
+        my $form = shift;
+        if ($form->c->cobrand->moniker eq 'peterborough') {
```
It would be better to have a `COBRAND_FEATURES` flag for the new feature rather than hardcoding the `peterborough` check everywhere. This would make it easier to roll out to other cobrands later and also means we can turn it on/off for Peterborough with a redeploy rather than code changes.

davea commented on perllib/FixMyStreet/Roles/CobrandBulkyWaste.pm:
```diff
@@ -552,10 +552,14 @@ sub bulky_reminders {
     my ($self, $report, $h, $params) = @_;
 
-    return unless $report->user->email;
+    return unless $report->user->email || $report->get_extra_field_value('bulky_text_updates');
```
This is inside a function called `_bulky_send_reminder_email` so shouldn't have anything to do with sending texts - can you factor this (and the below `call_hook` lines) into a new `_bulky_send_reminder_text` function which gets called independently?

davea commented on t/app/controller/waste_peterborough_bulky.t:
```diff
@@ -64,6 +66,7 @@ create_contact(
     { code => 'payment' },
     { code => 'payment_method' },
     { code => 'property_id' },
+    { code => 'bulky_text_updates' },
```
This is probably me not paying close enough attention but it took me ages of wondering why the feature wasn't working before I spotted this line and realised I also needed to add a `bulky_text_updates` hidden field to my existing `Bulky collection` category. Always helpful to please include notes for the reviewer when adding code that needs configuration/changes beyond checking out the branch.

davea commented on perllib/FixMyStreet/App/Form/Waste.pm:
```diff
@@ -55,17 +56,25 @@ sub validate {
-        unless $email->is_inactive || $email->value || ($is_staff_user && !$staff_provide_email);
-
+        unless $email->is_inactive || $email->value || ($is_staff_user && !$staff_provide_email)
+        || (FixMyStreet::SMS->parse_username($phone->value)->{may_be_mobile} && $cobrand eq 'peterborough');
```
This line seems to make the email address optional on the "About you" page if I provide a mobile number, but I can't complete a booking without an email address: 
<img width="548" alt="image" src="https://github.com/mysociety/fixmystreet/assets/4776/38db8122-041a-4afd-811c-7e2ba2d56df9">

MorayMySoc replied:
Fix to this in ba1e0155f9f0afe8ec29ac485f1102d9dad002d3 May need more discussion...

davea commented on perllib/FixMyStreet/App/Form/Waste/AboutYou.pm:
```diff
@@ -62,6 +62,14 @@ has_field email => (
     },
 );
 
+has_field extra_bulky_text_updates => (
```
The wording here isn't very clear, what about something like:
<img width="590" alt="image" src="https://github.com/mysociety/fixmystreet/assets/4776/2f33b2c4-7a87-447c-a2d6-5fbab41ca2e1">

davea reviewed (changes requested):
Functionally this is nearly there, but I've made a few comments about ways the code could be improved/tidied up a bit.

davea commented on perllib/FixMyStreet/App/Form/Waste/AboutYou.pm:
```diff
@@ -62,6 +62,14 @@ has_field email => (
 
+has_field extra_bulky_text_updates => (
+    type => 'Checkbox',
+    label => 'Bulky text updates',
```
I still think the inclusion of 'updates' here is misleading as the message aren't really updates - they're reminders.

MorayMySoc committed:
[Peterborough][WW] Rename extra_bulky_text_updates field

Renaming to be clearer about what it's for which is for
sending reminders about bulky waste collections which
have been ordered.

dracos committed:
[Peterborough] Bulky text updates/staff only.

davea committed:
Use separate GOV.UK config for updates on WasteWorks reports
